Refactor Node class to move Shape class usage to new derived TsNode

wconstab opened the pull request:
Begins the process of paring ir::Node down to the core features we want to share between backends, and adds infrastructure to let backends implement a derived, customized Node class.
- adds codegen support for backend derived class
- implements TsNode class
- moves lazy_tensors::Shape usages (primarily Node::shape(), Value::shape(), Output::shape() from ir::Node to TsNode 

Currently unfinished: 
- change how ir::Node hashing works, so backend Node passes node hash into ir::Node ctor and ir::Node doesn't have to decide whether to include or exclude Shape from hash
- update ir::Node ctor for aten shape/dtype instead of lazy_tensors::Shape

Note to any reviewers- go commit by commit, don’t review all the files one at a time.  I kept the commits separate so they are easier to review, and the ‘real’ changes separate from the ‘sed’ changes.  (there is a 'changes from _all commits_' dropdown in the 'Files Changed' tab.

wconstab committed:
Node Base [1/3]: Add Empty TsNode class

TsNode does nothing unique yet, just wraps Node.

This should not change any functionality,
just separates refactoring changes from more interesting ones later.

Modifies codegen scripts to accept customizable 'backend node' class,
so XLA backend can feed in their own.

Places that should use Node base:
- Graph traversals, APIs accepting a 'post_order' list of ops
- Debug Utils (should provide extension points for customization)
- Clone
- TensorMethods (which should also be going away for TS backend)

Places that should use the BackendNode  derived class:
- Lowerings
- Shape Inference
- IR classes (assuming they need specific functionality)

wconstab committed:
Node Base [2/3]: Bulk change ops to derive TsNode

More purely mechanical changes, this part kept separate since it is
strictly sed commands and touches so many files.

This is the equivalent for the handwritten op classes as the codegen
changes in Node Base [1/3].

find lazy_tensor_core/csrc/ops/ -name "*.h" | xargs sed -i 's/public Node/public TsNode/g'

find lazy_tensor_core/csrc/ops/ -name "*.h" | xargs sed -i 's/ir.h/ts_backend\/TsNode.h/g'

find lazy_tensor_core/csrc/ops/ -name "*.cpp" | xargs sed -i 's/: Node(/: TsNode(/g'

wconstab committed:
Node Base [3/3] Move shape() from Node to TsNode

wconstab committed:
Use unsafe dynamic_ptr_cast to access TsNode::SetShapeDefered (TODO refactor)

wconstab committed:
remove Shape from ir::Output [1/2]

- remove shape methods from Output struct
- add helper to TsNode for casting Output::node to a TsNode* and
  accessing TsNode::Shape

wconstab committed:
remove Shape from ir::Output [2/2]

- apply GetShapeFromTsOutput helper to all usages in
  ts_node_lowering.cpp

wconstab committed:
Remove shape() from Value [1/2]

- add GetShapeFromTsValue helper for getting TsNode::shape() from
  a Value that holds a TsNode

wconstab committed:
Remove shape() from Value [2/2]

- bulk change to use GetShapeFromTsOutput helper in all ops classes

sed -i -r
  's/([a-zA-Z]+[a-zA-Z0-9_]*)\.shape\(\)/GetShapeFromTsValue\(\1\)/g'
  lazy_tensor_core/csrc/ops/*.cpp

- manual fixes to undo wrong sed changes in
  ops/amp_foreach_non_finite_check_and_unscale.cpp
  ops/index_ops.cpp
- and manually added missing spots in tensor_methods.cpp and tensor.cpp

wconstab committed:
Remove Node::shape

- since Value and Output stopped using it, and other uses have migrated
  to TsNode::shape, this is mostly trivial
- deal with ToString and ir_dump utils with workarounds for now
- add a GetShapeFromTsNode helper to deal with one usage in index_ops

wconstab committed:
Remove other traces of Shape class from Node

wconstab committed:
Use TsNodeSetShapeDeferred instead of unsafe dynamic_cast

alanwaketan reviewed (commented):
Not sure what's the intent to remove shape from the base Node class. If the concern is about different backend wants to use different Shape types, then we can easily use a typedef/using combined with some compile time flags to solve the problem.
A rough example:
```
#ifdef TORCH_SCRIPT_BACKEND
using Shape = vector<vector<int>>;
#endif
```

wconstab commented:
> If the concern is about different backend wants to use different Shape types, then we can easily use a typedef/using combined with some compile time flags to solve the problem. 

Lots of clients of (TsNode::shape(), now) use Shape class's rich API directly, and they would not be happy if Shape were suddenly changed under their feet to be a vector.

> Not sure what's the intent to remove shape from the base Node class.
1. Remove nonstandard types (we want to use the same types as native pytorch, not make new type clases)
2. Provide a reference implementation for XLA: they will keep using xla::Shape in their codebase, so our design needs a way to have XlaNode derive from Node and add Shape class back in.  I felt it was easier to prototype TsNode than prototype XlaNode, but it shows the same functionality.  Xla team would have to build the real XlaNode, following this TsNode example.
3. Help decouple LT Core from LT ts_backend:  We still have a fair amount of work to get ts_backend cleaned up, but it's less worrisome if some nonstandard types linger there than in core.  In order to start landing core pieces asap, we need to start separating the two and understand which parts really belong in core vs backend.

Will we add shape() back to Node base, but using standard aten types?  Probably, but it might not look the same as the current shape() api.  We need to sort out the requirements of full- and bounded- dynamic shape backends, and determine if we're still supporting any static-shape backends.  We'll definitely be changing how shape interacts with node hash.

wconstab commented on lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/TsNode.cpp:
```diff
@@ -0,0 +1,24 @@
+namespace torch_lazy_tensors {
+namespace ir {
+
+TsNode::TsNode(OpKind op, OpList operands, lazy_tensors::Shape shape,
```
We can change that.  But to be clear, is my change introducing a new problem or it's the same as the original code?  If not a new problem, better to fix in a separate PR, but I still support improving it.

alanwaketan replied:
It's not a new problem. The original Node class implemented it in this way. That's why I suggest a follow-up.

alanwaketan commented:
@wconstab Thanks for writing a very detailed explanation. So if I understand it correctly, we want a 'Shape' type in the base Node class but we don't know what the ideal 'Shape' type yet. Then I think my proposal above can help.

One point I forgot to add, I support this TSNode : Node inheritance change. I was kinda thinking that you can still keep the shape related interfaces in the base node while doing the implementation in the TSNode. For the type discrepancy,  you can then use the above typedef techniques to resolve all your down-casting problems. Overall, the base Node class doesn't need to know what the actual 'Shape' type is.

If you are worried whether it's good for the 'core' directory to be polluted by some backend compile time flags or not, I'd say it should be fine. After all, you have to build it against a backend. As long as there aren't too many of these, it should be good. Also, we have a good intent here given 'Shape' is one of the inseparable parts of a Node.

JackCaoG commented on lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/TsNode.cpp:
```diff
@@ -79,5 +86,18 @@ lazy_tensors::Shape TsNode::TsGetOpShape(
 
+std::string TsNode::ToString() const {
+  std::stringstream ss;
+  ss << shape() << " " << op();
```
The only thing that is backend dependent here is `shape`. Would it be better to leave `ToString` in the `Node` and have each backend Node to implement a `ShapeToString()`? 

The downside of this approach I suggested would be
1. Each backend might want to implement `ToString` differently
2. Passing string as value might be expensive

The upside is backend doesn't need to manually update its `ToString` method when there is a new change in the `Node` base class (For example `aten::shape` you want to add). You also don't need to do a Cast in `Node` to get the backend shape for hash now.

JackCaoG replied:
I think Backend can just override ToString entirely if they want, we can skip `ShapeToString` in this case.

Krovatkin replied:
if we can get away putting actual fields in the base class by overriding `ToString` we should probably do that

JackCaoG reviewed (commented):
Thanks @wconstab . I think you added `GetShapeFromTs***` because each backend will use a different shape class and it can't be expressed in the `Node` base class. Is that right?

I think when `Node` doesn't have any dependency to the `LazyTensor::` types, pytorch/xla can start implementing the `XlaNode` class.

wconstab commented:
> Thanks @wconstab . I think you added `GetShapeFromTs***` because each backend will use a different shape class and it can't be expressed in the `Node` base class. Is that right?

Yes, I added these for the reason you assumed.  But I was chatting with @alanwaketan and we came up with a possibly cleaner idea.  The `GetShapeFromTs***` helpers are basically there to do downcasting for you.  We might be able to avoid this.  My idea (untested yet) is to make the `LazyTensor` class a template class with type NodeT, and use static compiler check that NodeT derives from Node.  In this way, LazyTensor can be shared by Xla and TS, but the NodeT can be a specific XlaNode or TsNode.  LazyTensor methods themselves can be implemented strictly to Node API.  But when presenting a node from LazyTensor to Lower* functions, it can be presented as real XlaNode type, and Lower* functions can directly access XlaShape() field without a downcast.  I don't know how to do the static check but I hope there is a way.

wconstab commented:
> Not sure what's the intent to remove shape from the base Node class. If the concern is about different backend wants to use different Shape types, then we can easily use a typedef/using combined with some compile time flags to solve the problem. A rough example:
> 
> ```
> #ifdef TORCH_SCRIPT_BACKEND
> using Shape = vector<vector<int>>;
> #endif
> ```

@alanwaketan We discussed offline, but recapping here for posterity.

My intent actually _is_ to remove the lazy_tensors::Shape (aka xla::Shape) class from Node.  This is to 'purify' Node and keep it in terms of core PyTorch types.  In this PR, I try to pave a way for XlaNode (derived from Node) to add back xla::Shape for its own usage only.  Code using xla::Shape would be limited to XLA backend/lowerings, but code in lazy core (e.g. LazyTensor class, other utils) should use Node API and not xla::Shape.

Regarding typedef suggestion, as we discussed it wouldn't work to typedef a Shape type differently for different backends, unless all the call sites of Shape used the same Shape API.  Since the xla::Shape API is rich and not similar to the pytorch core (vector<int>) approach, the typedef wouldn't directly help.  It would help if we decided to make our own 'torch::Shape' class, but, that's not the proposal.

I am proposing to add aten's shape/dtype into Node base, however.  I haven't decided whether to call their accessors `aten_shape` or just `shape` (in the latter case, forcing XlaNode and XLA code to rename node->shape() to node->xlaShape() or something).
@JackCaoG would you be opposed if I make Node::shape() return the pytorch type and require xla code to access an explicit xlaShape() method?

Krovatkin commented on lazy_tensor_core/lazy_tensor_core/csrc/op_by_op_executor.cpp:
```diff
@@ -52,7 +52,7 @@ lazy_tensors::hash_t ComputeNodeKey(
   }
   key = lazy_tensors::util::HashCombine(
-      key, lazy_tensors::util::ShapeHash(node->shape()));
+      key, lazy_tensors::util::ShapeHash(GetShapeFromTsNode(*node)));
```
each backend will override their hashing or will we standardize on what `shape()` from the base class return?

Krovatkin commented on lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/TsNode.cpp:
```diff
@@ -0,0 +1,24 @@
+namespace ir {
+
+TsNode::TsNode(OpKind op, OpList operands, lazy_tensors::Shape shape,
+               size_t num_outputs, lazy_tensors::hash_t hash_seed)
```
it's so weird an IR node needs to worry about `hash_seed` @wconstab  @alanwaketan  do you guys know if we ever pass anything different from the default value in there?

wconstab replied:
oh, timely comment! I happen to have just renamed all of this on my local copy.  I agree, it was a bit of a misnomer.  In this case, the 'seed' is really a partial hash passed in by the leaf class, which hashes over any scalars/constants provided to the leaf ctor.

Krovatkin commented on lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/TsNode.cpp:
```diff
@@ -1,24 +1,69 @@
+  return cache;
+}
+
+lazy_tensors::Shape TsNode::TsGetOpShape(
```
nit: the name is a bit funny. It's really used for setting shapes rather than getting them. At first, I got confused between the two `shape` and `TsGetOpShape` but just looking at the uses made it clear what the difference was

wconstab committed:
Remove Shape from Node, Compute Hash in TsNode

- rename hash_ to dag_hash_ to clarify its purpose
- change Node ctor to accept node_hash and dag_hash computed by
  BackendNode
- update some cases in ts_node_lowering that should have been using
  dynamic-cast helpers

JackCaoG commented:
> > Not sure what's the intent to remove shape from the base Node class. If the concern is about different backend wants to use different Shape types, then we can easily use a typedef/using combined with some compile time flags to solve the problem. A rough example:
> > ```
> > #ifdef TORCH_SCRIPT_BACKEND
> > using Shape = vector<vector<int>>;
> > #endif
> > ```
> 
> @alanwaketan We discussed offline, but recapping here for posterity.
> 
> My intent actually _is_ to remove the lazy_tensors::Shape (aka xla::Shape) class from Node. This is to 'purify' Node and keep it in terms of core PyTorch types. In this PR, I try to pave a way for XlaNode (derived from Node) to add back xla::Shape for its own usage only. Code using xla::Shape would be limited to XLA backend/lowerings, but code in lazy core (e.g. LazyTensor class, other utils) should use Node API and not xla::Shape.
> 
> Regarding typedef suggestion, as we discussed it wouldn't work to typedef a Shape type differently for different backends, unless all the call sites of Shape used the same Shape API. Since the xla::Shape API is rich and not similar to the pytorch core (vector) approach, the typedef wouldn't directly help. It would help if we decided to make our own 'torch::Shape' class, but, that's not the proposal.
> 
> I am proposing to add aten's shape/dtype into Node base, however. I haven't decided whether to call their accessors `aten_shape` or just `shape` (in the latter case, forcing XlaNode and XLA code to rename node->shape() to node->xlaShape() or something). @JackCaoG would you be opposed if I make Node::shape() return the pytorch type and require xla code to access an explicit xlaShape() method?

Since Node base class will only carry the aten shape going forward, I am fine with `shape()` returns an aten shape. XLATensor also have `logical_element_type ` which will return a `at::ScalarType`, if we ended up decide to make `shape == aten_shape`, I will rename this `logical_element_type` to `type` too.

wconstab committed:
Remove unused APIs from NodeBase (uses, Replace*)

- uses(), ReplaceOperand, ReplaceAllUsesWith removed
- appears to be leftover from previous version? Current code doens't
  modify a node, probably switched to using 'Clone' instead

wconstab committed:
Move operands() related stuff to TsNode

- Later, we should avoid storing these vectors in the node, and instead
  use the codegenerated classes to directly implement operands() from
their underlying operand fields.  For now, since there are too many
non-codegen IR classes in play, we keep it this way in TsNode as a
convenience.

wconstab committed:
Move aten shapes/sizes to codegen nodes

- rolling back a bit from adding these fields to Node base
- it's a nice idea to have them in Node base, but (a) not necessary yet,
  and (b) we're not ready to implement it in a consistent way for
  non-codegen nodes
- non-codegen nodes don't all have their shapes initialized at ctor
  time, meaning that even if we infer the at_shapes fields from the
  lazy_tensors::Shape fields once they are set, the at_shapes API could be
  called at a time when the fields aren't yet initialized.

wconstab committed:
Fix ToString

wconstab committed:
Fix format

wconstab committed:
Change OpList to c10::ArrayRef

- have to wrap many places in at::ArrayRef explicitly for a reason I don't
  understand

alanwaketan reviewed (commented):
I like the overall design and the motivations behind it. Just a few comments need to be addressed before landing.

wconstab committed:
Address PR feedback

wconstab committed:
Change NodeCast to const

alanwaketan reviewed (approved):
LGTM. It's not easy to make such a huge change. Good job!
